File: MonteCarlo/MC_Integration_1.cc
void MC_Integration_1(void){

  //Declare some useful variables here
  TF1* f1 = new TF1("f1","3*x^3 - x",0,1);

  int nTrials = 2000;
  vector<double> trials;
  double nSamples = 50;

  double mean = 0;
  double variance = 0;

  TRandom r(1524);
    TH1D* hi = new TH1D("Integrals","Integrals",100,0,0.5);

  //run several trials to get a sense for mean and RMS
  for(int t = 0; t<nTrials; t+=1){    

    double integral = 0;
    //For each trial, use the same number of samples
    for(int b = 1; b<=nSamples; b+=1){
      double xi = r.Uniform();
      double fi = f1->Eval(xi)/nSamples;  //Why divide by nSamples?
      integral += fi;
    }

    cout << "Integral: " << integral << endl;
    hi->Fill(integral);
    
    trials.push_back(integral);
    mean += integral;
  }
  mean /= nTrials;

  cout << "Mean: " << mean << endl;

  // Calculate average (x-mu)^2
  for(int t = 0; t<nTrials; t+=1){    
    variance += TMath::Power(trials[t]-mean,2);
  }
  
  variance /= nTrials;
  variance = sqrt(variance);  //variance is (std dev)^2
  
  cout << "Std Dev: " << variance << endl;
 TCanvas* c1 = new TCanvas("c1","c1");
  hi->Draw();
}

File: MonteCarlo/MC_Integration_1.py
import numpy as np
import matplotlib.pyplot as plt

def myFunction(rand):
    return 3*pow(rand,3) - rand


#Declare some useful variables here
nTrials = 20
nSamples = 50
mean = 0
variance = 0

trials = []

#run several trials to get a sense for mean and RMS
for t in range(0,nTrials):
    integral = 0

    #For each trial, use the same number of samples
    for b in range(0,nSamples):
        xi = np.random.random()    #Uniform in [0.0,1.0)
        fi = myFunction(xi)/nSamples  #Why divide by nSamples?
        integral += fi

    print("Integral:  %.5f" % integral)
    
    trials.append(integral)
    mean += integral
  
mean /= nTrials

print("Mean: %.5f" % mean)

# Calculate average (x-mu)^2
for t in range(0,nTrials):
    variance += pow(trials[t]-mean,2)

variance /= nTrials
variance = np.sqrt(variance)  #variance is (std dev)^2
  
print("Std Dev: %.5f" % variance)

#Make a plot of these values!
plt.hist(trials, density=False, bins=30)  # `density=True` would normalize to 1
plt.ylabel('Frequency')
plt.xlabel('Integral');
plt.show()


